### PR TITLE
Bug #320 - JSR validation language

### PR DESCRIPTION
**Pull request by thibaultmeyer**
This pull request fix issue on the JSR303 error messages locale (Bug  https://github.com/ninjaframework/ninja/issues/320). Before this patch, JSR used the locale of the running server : _Locale.getDefault()_. Now, the locale is determined by the value of the request's header _Accept-Language_. The determined locale is only apply to the current request (not system wild).

### CODE REVIEW

File: ninja-core/src/main/java/ninja/validation/Validators.java [hunk 1]
```diff
@@ -16,40 +16,89 @@
...
+            this.validate(value, field, validation, null);
+        }
+
+        @Override
+        public void validate(Object value, String field, Validation validation, ninja.Context context) {
+            if (value != null) {
+                final ValidatorFactory validatorFactory = javax.validation.Validation.buildDefaultValidatorFactory();
+                final javax.validation.Validator validator = validatorFactory.getValidator();
+                final Set<javax.validation.ConstraintViolation<Object>> violations = validator.validate(value);
+                final Locale localeToUse = (context == null || context.getAcceptLanguage() == null) ? Locale.getDefault() : Locale.forLanguageTag(context.getAcceptLanguage().split(",", 2)[0]);
```
Comment (raphaelbauer) [inline ninja-core/src/main/java/ninja/validation/Validators.java]: The perfect way is to use @Inject Lang and then lang.getLanguage(context, Optional.absent())

The language can not only be in the accept header, but also inside a cookie that forces Ninja to use a certain language. Lang supports that.

You can inject Lang into JSRValidator's constructor.

File: ninja-core/src/main/java/ninja/validation/Validator.java [hunk 2]
```diff
@@ -32,9 +34,24 @@
...
+     * @param value
+     *            The value, may be null
+     * @param field
+     *            The name of the field being validated, if applicable
+     * @param validation
+     *            The validation context
+     * @param context
+     *            The Ninja request context
+     */
+    void validate(T value, String field, Validation validation, Context context);
```
Comment (raphaelbauer) [inline ninja-core/src/main/java/ninja/validation/Validator.java]: Having validation AND context seems redundant - could we just have context (which got context.getValidation?).

That's of course not really backwards compatible - do you think we can make that? I actually don't have big problems with breaking backwards compatibility in a - say Ninja 4.1 release.

**Comment by raphaelbauer** [comment]
I added some small comments. You solution looks almost perfect :)

We also need a testcase for the final implementation - and also an entry in changelog.md.

File: ninja-core/src/main/java/ninja/validation/Validator.java (hunk 2 above)
Comment (thibaultmeyer) [inline ninja-core/src/main/java/ninja/validation/Validator.java]: no problem, I'll see what I can do to remove the variable validation tonight

File: ninja-core/src/main/java/ninja/validation/Validators.java (hunk 1 above)
Comment (thibaultmeyer) [inline ninja-core/src/main/java/ninja/validation/Validators.java]: Nice, i will try to use the Lang class. thank you for the tip

**Comment by thibaultmeyer** [comment]
I finish to change my code to use the Lang class and I begin writing unit tests for this pull request.

**Comment by thibaultmeyer** [comment]
Using @Inject Lang in the JSRValidator constructor doesn't work. Maven fail during execution of a "none related" tests (SwissKnifeTest).  If I remove the newly added construction from the JSRValidator class, the Maven process run with success.

maybe I do not use it correctly ?

###### JSRValidator

```
public static class JSRValidator implements Validator<Object> {

    private final Lang requestLanguage;

    @Inject
    public JSRValidator(Lang requestLanguage) {
        this.requestLanguage = requestLanguage;
    }

    @Override
    public void validate(Object value, String field, Context context) {
        if (value != null) {
            final ValidatorFactory validatorFactory = javax.validation.Validation.buildDefaultValidatorFactory();
            final javax.validation.Validator validator = validatorFactory.getValidator();
            final Set<javax.validation.ConstraintViolation<Object>> violations = validator.validate(value);
            final Locale localeToUse = (context.getAcceptLanguage() == null) ? Locale.getDefault() : Locale.forLanguageTag(context.getAcceptLanguage().split(",", 2)[0]);
            final Validation validation = context.getValidation();

            for (javax.validation.ConstraintViolation<Object> violation : violations) {
                final String violationMessage = validatorFactory.getMessageInterpolator().interpolate(
                        violation.getMessageTemplate(),
                        new NinjaContextMsgInterpolator(value, violation.getConstraintDescriptor()),
                        localeToUse
                );
                final ConstraintViolation constraintViolation = ConstraintViolation.create(violationMessage, violation.getInvalidValue());
                validation.addBeanViolation(new FieldViolation(violation.getPropertyPath().toString(), constraintViolation));
            }
        }
    }

    @Override
    public Class<Object> getValidatedType() {
        return Object.class;
    }
}
```

###### Maven

```
...
Running ninja.utils.ResultHandlerTest
Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.013 sec
Running ninja.utils.SecretGeneratorTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running ninja.utils.SwissKnifeTest

Running ninja.utils.SwissKnifeTest
22:02:39.014 [main] ERROR ninja.utils.SwissKnife - No converter found to convert class java.lang.StringBuffer. Returning null. You may want to extend the class.
22:02:39.015 [main] ERROR ninja.utils.SwissKnife - Cannot convert from java.lang.String to java.lang.Integer. Conversion failed with null
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_40]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_40]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_40]
        at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_40]
        at ninja.utils.SwissKnife.convert(SwissKnife.java:159) ~[classes/:na]
        at ninja.utils.SwissKnifeTest.testConvert(SwissKnifeTest.java:63) [test-classes/:na]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_40]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_40]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_40]
        at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_40]
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47) [junit-4.11.jar:na]
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.11.jar:na]
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44) [junit-4.11.jar:na]
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) [junit-4.11.jar:na]
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271) [junit-4.11.jar:na]
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70) [junit-4.11.jar:na]
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50) [junit-4.11.jar:na]
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238) [junit-4.11.jar:na]
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63) [junit-4.11.jar:na]
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236) [junit-4.11.jar:na]
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53) [junit-4.11.jar:na]
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229) [junit-4.11.jar:na]
        at org.junit.runners.ParentRunner.run(ParentRunner.java:309) [junit-4.11.jar:na]
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252) [surefire-junit4-2.12.4.jar:2.12.4]
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141) [surefire-junit4-2.12.4.jar:2.12.4]
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112) [surefire-junit4-2.12.4.jar:2.12.4]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_40]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_40]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_40]
        at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_40]
        at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189) [surefire-api-2.12.4.jar:2.12.4]
        at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165) [surefire-booter-2.12.4.jar:2.12.4]
        at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85) [surefire-booter-2.12.4.jar:2.12.4]
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115) [surefire-booter-2.12.4.jar:2.12.4]
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75) [surefire-booter-2.12.4.jar:2.12.4]
        Caused by: java.lang.NumberFormatException: For input string: "aString"
        at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[na:1.8.0_40]
        at java.lang.Integer.parseInt(Integer.java:580) ~[na:1.8.0_40]
        at java.lang.Integer.valueOf(Integer.java:766) ~[na:1.8.0_40]
        at ninja.utils.SwissKnife$Converter.toInteger(SwissKnife.java:180) ~[classes/:na]
        ... 35 common frames omitted
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 sec
Running ninja.utils.TimeUtilTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 sec

Results :

Tests in error:
(..)lidationWithNullObject(ninja.params.ControllerMethodInvokerTest): Guice configuration errors:
(..)lidationFailedWithThreeFields(ninja.params.ControllerMethodInvokerTest): Guice configuration errors:
(..)lidationFailedLength(ninja.params.ControllerMethodInvokerTest): Guice configuration errors:
(..)lidationFailedWithTwoAnnotations(ninja.params.ControllerMethodInvokerTest): Guice configuration errors:
(..)lidationPassed(ninja.params.ControllerMethodInvokerTest): Guice configuration errors:
(..)lidationFailedRange(ninja.params.ControllerMethodInvokerTest): Guice configuration errors:
(..)lidationFailedRegex(ninja.params.ControllerMethodInvokerTest): Guice configuration errors:

Tests run: 385, Failures: 0, Errors: 7, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] ninja .............................................. SUCCESS [  0.482 s]
[INFO] ninja-core ......................................... FAILURE [ 16.674 s]
[INFO] ninja-metrics ...................................... SKIPPED
[INFO] ninja-metrics-graphite ............................. SKIPPED
```

**Comment by raphaelbauer** [comment]
Wohooo! That looks all good :)
Actually the error you spotted (ninja.utils.SwissKnife) is just a test to see if the SwissKnife throws an error. So that's absolutely okay.

Your problem comes from ninja.params.ControllerMethodInvokerTest (Tests in error).

ControllerMethodInvokerTest is a giant test for all Validators in Ninja (Yes - it's possibly too big and should be refactored). It has a class ControllerMethodInvoker and a configure method to create an injector. Because Lang is something new you have to tell the injector to bin Lang to something...  bind(Lang.class).toInstance(...)

Not sure how you should do that. Maybe just mock Lang and check in the JSRValidation if stuff works :) That's then up to you!

**Comment by thibaultmeyer** [comment]
Now Lang::getLanguage() is used  to determine the language to use for the JSR303 error message. Unit tests has been fixed to use the Lang instance via Mock.

**Comment by raphaelbauer** [comment]
This is AWESOME! Very clean and good work!

**Comment by raphaelbauer** [comment]
I added a comment to the changelog containing your implementation.

**Pull request merged**